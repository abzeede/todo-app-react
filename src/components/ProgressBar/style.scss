.progress {
  display: flex;
  flex-direction: row;
  align-items: flex-end;
  width: 100%;

  &__text {
    font-size: 0.9em;
    margin-right: 0.5em;
    color: white;
  }

  &__bar {
    width: 100%;
    height: 10px;
    background-color: #fff;
    border-radius: 3px;
    overflow: hidden;

    &--active {
      background-color: #38b549;
      height: 100%;
      transition: 0.2s ease-in-out;
    }
  }
}
