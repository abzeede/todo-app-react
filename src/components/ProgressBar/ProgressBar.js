import React from 'react'
import { arrayOf, shape, number, string, bool } from 'prop-types'
import './style.scss'

const ProgressBar = ({ items, showPercent }) => {
  const calculateProgress = () => {
    const itemCount = items.length
    if (itemCount === 0) {
      return 0
    }

    return 100 - Math.floor((100 * items.filter(item => !item.done).length) / itemCount)
  }
  const progress = calculateProgress()

  return (
    <div className="progress">
      {showPercent && <div className="progress__text">{progress}%</div>}
      <div className="progress__bar">
        <div className="progress__bar--active" style={{ width: `${progress}%` }} />
      </div>
    </div>
  )
}

ProgressBar.propTypes = {
  items: arrayOf(
    shape({
      id: number,
      title: string,
      done: bool,
    }),
  ).isRequired,
  showPercent: bool,
}

ProgressBar.defaultProps = {
  showPercent: false,
}

export default ProgressBar
