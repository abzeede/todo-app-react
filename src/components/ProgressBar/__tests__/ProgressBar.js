import React from 'react'
import { mount } from 'enzyme'
import toJson from 'enzyme-to-json'
import ProgressBar from '../ProgressBar'

describe('ProgressBar', () => {
  it('should render properly', () => {
    const items = [
      { id: 1, detail: 'todo 1', done: false },
      { id: 2, detail: 'todo 2', done: true },
      { id: 3, detail: 'todo 3', done: false },
      { id: 4, detail: 'todo 4', done: false },
    ]

    const wrapperShowPercent = mount(<ProgressBar items={items} showPercent />)
    const wrapperNotShowPercent = mount(<ProgressBar items={items} showPercent={false} />)

    expect(wrapperShowPercent.find('.progress__text').text()).toBe('25%')
    expect(wrapperShowPercent.find('.progress__bar--active').props().style.width).toBe('25%')
    expect(toJson(wrapperShowPercent)).toMatchSnapshot()

    expect(wrapperNotShowPercent.find('.progress__text')).toHaveLength(0)
    expect(toJson(wrapperNotShowPercent)).toMatchSnapshot()
  })
})
