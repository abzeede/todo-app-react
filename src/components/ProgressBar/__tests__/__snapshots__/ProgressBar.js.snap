// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar should render properly 1`] = `
<ProgressBar
  items={
    Array [
      Object {
        "detail": "todo 1",
        "done": false,
        "id": 1,
      },
      Object {
        "detail": "todo 2",
        "done": true,
        "id": 2,
      },
      Object {
        "detail": "todo 3",
        "done": false,
        "id": 3,
      },
      Object {
        "detail": "todo 4",
        "done": false,
        "id": 4,
      },
    ]
  }
  showPercent={true}
>
  <div
    className="progress"
  >
    <div
      className="progress__text"
    >
      25
      %
    </div>
    <div
      className="progress__bar"
    >
      <div
        className="progress__bar--active"
        style={
          Object {
            "width": "25%",
          }
        }
      />
    </div>
  </div>
</ProgressBar>
`;

exports[`ProgressBar should render properly 2`] = `
<ProgressBar
  items={
    Array [
      Object {
        "detail": "todo 1",
        "done": false,
        "id": 1,
      },
      Object {
        "detail": "todo 2",
        "done": true,
        "id": 2,
      },
      Object {
        "detail": "todo 3",
        "done": false,
        "id": 3,
      },
      Object {
        "detail": "todo 4",
        "done": false,
        "id": 4,
      },
    ]
  }
  showPercent={false}
>
  <div
    className="progress"
  >
    <div
      className="progress__bar"
    >
      <div
        className="progress__bar--active"
        style={
          Object {
            "width": "25%",
          }
        }
      />
    </div>
  </div>
</ProgressBar>
`;
